### ISSUE TITLE
{textsize} doesn't quite revert to original size

### ISSUE THREAD

**Issue by gwyndaf** [ChordPro/chordpro#262]
```
{title: textsize}
Here's a [D]regular line
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
{textsize: 50%}
This should be [G]small
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
{textsize}
This should revert to [E]regular size
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
```
Produces this attached output:
[testsize.pdf](https://github.com/ChordPro/chordpro/files/10287836/testsize.pdf)

I'd expect the third line of MMM to be the same length (i.e. typesize) as the first, but it's a bit shorter, so hasn't quite reverted to the original size.

What seems to affect it is this part of my config, and the problem goes away if I comment this out:
```
{
    "pdf" : {
      "fonts" : {
        "text" : {
          "description" : "sans 13"
        },
      },
  },
}
```
I think maybe what's happening is that `{textsize}` isn't reverting to the previous size (13pt) as set in my own config, but to something like 12pt, which I suspect is maybe coming from the default (built-in) config.

**Comment by sciurius** [comment]
The short answer is that your config is merged into the default config, so the the **full** font specification looks like this:
````
{
    "pdf" : {
      "fonts" : {
        "text" : {
          "name" : "Times-Roman",
          "size" : 12,
          "description" : "sans 13"
        },
      },
  },
}
````
So the short workaround is to add `"size":13` to your config.

I agree that this is not intuitive, I'll take a look.

**Comment by sciurius** [comment]
Given this:
````
[A]Default font
{textfont Times 20}
[A]This is Times, size 20
{textfont Times-Roman}
[A]This is Times, but what size?
````
What would you expect?

**Comment by gwyndaf** [comment]
Intuitively, I'd expect the first directive to:

1.  Set font face to `Times`, if it's recognised. I don't think it's strictly one of ChordPro's built-it Postscript font names, but might exist as a system font (e.g. on MacOS); or could have been defined (as description) in a config file.
2. Set font size to `20`.


I think I'd expect the second directive would 

3. Change the font face to `Times-Roman` (built-in name, with no bold or italic).
4. Leave the size unchanged, i.e. `20`.

Point 4. probably depends if the second directive is interpreted as only changing the font _face_, which I think is the clearest implication.

I can see that the absence of a size could potentially imply `{textsize}` (i.e. reset to previous/original size), but I think that's counter-intuitive, and maybe better requiring an explicit `{textsize}` to reset to former size.

**Comment by sciurius** [comment]
So the semantics would be that a size-changing `textfont` implies `textsize`.

    {textfont Times 20}

behaves as if written as

    {textfont Times 20}
    {textsize 20}

A subsequent `{textfont}` will reset the font face, but not the size.

**Comment by gwyndaf** [comment]
A similar test for font weight/style seems to behave as expected:
```
Here's a [D]regular line
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
{textfont: sans bold italic}
Expect this to be [G]bold italic
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
{textfont: sans italic}
Expect this to be [G]italic
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
{textfont}
Expect this to revert to [E]previous bold italic style (not config default)
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM
```

This seems to make sense, because I think the scope of `{textfont}` is face, weight and style/slant (plus optionally size). Issue seems only to arise because of the overlap created by the overlap (size) between `{textfont}` and `{textsize}` 

I guess the same issue applies to other font/size directives like `{chordfont}` and `{chordsize}`, but I'll set up a test to find out... :)

**Comment by gwyndaf** [comment]
I can confirm that same issue exists with `{chordsize}` and `{tabsize}` i.e. directive with no value applies built-in defaults, not reverting value from previous directives, unless `size` value specified in config.

**Comment by sciurius** [comment]
What happens at the start of each song is something similar to:

1. Save the current config.
2. Initialize the fonts
3. ...

Then `{textfont}` etc. are stacked. When the stack is empty, it defaults to the saved config.
This has worked in the past, but with font descriptors is it possible to change the font size as well. So when the stack is empty, the default should not come from the saved config (after step 1), but from the config **after** font initialisation (after step 2 above).

So what I will do:
1. If the `textfont` value ends with a number, use that as an explicit `textsize`.
2. When the stack is empty, default as described above.

**Comment by gwyndaf** [comment]
That seems sensible.

To clarify the config side (i.e. fallback once stack is empty): will a number at the end of a `<item>.description` value (e.g. `sans bold 13`) be interpreted as a `size` value, or will `size` still need to be explicit, as per your workaround?

**Comment by sciurius** [comment]
No, you won't have to add `"size": 13` to the config. It should default to the config **after** font initialisation.

**Comment by sciurius** [comment]
Can you try 5.990_010?

**Comment by gwyndaf** [comment]
That seems to fix it. Certainly in the simple test and one-off song samples.

Something isn't happening quite right (i.e. still as previous behaviour) when I use it in my broader songbook process. It's probably down to my own configs and/or script, but I'll need to investigate that further...

**Comment by sciurius** [comment]
There may have been a small oversight when multiple songs are involved. Can you try 5.990_011?

**Comment by gwyndaf** [comment]
Works with one collection, but another still has the previous behaviour. Seems like similar results to 4.990_010.

I think the cause might be an earlier song in the second collection, with a `{textsize: 92%}` directive (applying to the whole song): if I remove that, it behaves as expected.

**Comment by sciurius** [comment]
Can you boil it down to a smaller file with just 2-3 songs?

**Comment by gwyndaf** [comment]
I've trimmed it back to this sample: 
[test_book.cho.txt](https://github.com/ChordPro/chordpro/files/10298732/test_book.cho.txt)
In the first song, a single `{textsize: 60%}` applies to the whole song; in the second, a `{textsize: 70%} ... {textsize}` pair adjusts only one line, then I'd expect size to reset to config defaults.

I think I can now see what's happening: the final `{textsize}` is reverting to the value carried over from the previous song. Originally that was 92% in the first, so didn't look much different from the 12pt built-in default when the second song reverted to it, but is clearer with the more extreme 60%, as per the output:
[test_book.cho.pdf](https://github.com/ChordPro/chordpro/files/10298744/test_book.cho.pdf)

I'm assuming using a `{new_song}` separator within a single file leads to the same behaviour as a `--filelist` of song files.

Personally, I think it makes sense that each new song in a collection brings a reset to config defaults (not previous song), which I think has been former behaviour.

Rationale: current behaviour makes some sense if songs only exist in the context of a particular collection and sequence; however, with a repository of song files, which can be combined in many ways with `--filelist` it's impossible to anticipate what order they might finally be processed in.

**Comment by sciurius** [comment]
Well done. Indeed, every new song should start with exactly the same config.
Please try 5.990_012.

**Comment by gwyndaf** [comment]
Fantastic. That works with the test sample, and also seems to be working in full-scale songbook production :)
